### PR TITLE
Add basic Helm 3 support (with ListReleases only)

### PR DESCRIPTION
**Pull request by SimonAlling**
### Description of the change

It implements `ListReleases` and `ListAllReleases` using the Helm 3 API.

**Note:** `useHelm3` must be set to `true` in `values.yaml` for the changes introduced here to have any effect.

### Benefits

We should get one step closer to Helm 3 support.

### Possible drawbacks

There are no tests. (Hello, @absoludity!)

### Applicable issues

This PR emerged from the discussion in #1309.

### Additional information

We went with the name **Kubeops** this time (instead of Helmer).

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+func appOverviewFromRelease(r *release.Release) proxy.AppOverview {
+	return proxy.AppOverview{
+		ReleaseName: r.Name,
+		Version:     strconv.Itoa(r.Version),
+		Icon:        r.Chart.Metadata.Icon,
+		Namespace:   r.Namespace,
+		Status:      r.Info.Status.String(),
+	}
+}
```
Comment (absoludity) [inline pkg/agent/agent.go]: Great, I'll have a look at testing this and the handler tomorrow. Thanks!

File: chart/kubeapps/templates/kubeops-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  template:
+    metadata:
+      labels:
+        app: {{ template "kubeapps.kubeops.fullname" . }}
+        release: {{ .Release.Name }}
+    spec:
+      serviceAccountName: {{ template "kubeapps.kubeops.fullname" . }}
+      # Increase termination timeout to let remaining operations to finish before killing the pods
+      # This is because new releases/upgrades/deletions are synchronous operations
+      terminationGracePeriodSeconds: 300
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply chart/kubeapps/templates/kubeops-deployment.yaml]: I don't know tbh; this file is basically just `tiller-proxy-deployment.yaml` adapted for Kubeops.

File: chart/kubeapps/templates/kubeops-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        - --user-agent-comment=kubeapps/{{ .Chart.AppVersion }}
+        - --chartsvc-url=http://{{ template "kubeapps.chartsvc.fullname" . }}:{{ .Values.chartsvc.service.port }}
+        ports:
+        - name: http
+          containerPort: {{ .Values.kubeops.service.port }}
+        env:
+        - name: POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply chart/kubeapps/templates/kubeops-deployment.yaml]: I see. Shouldn't be any problem.

File: cmd/apprepository-controller/pkg/client/clientset/versioned/typed/apprepository/v1alpha1/apprepository_client.go [hunk 4]
```diff
@@ -71,7 +70,7 @@ func setConfigDefaults(config *rest.Config) error {
 	gv := v1alpha1.SchemeGroupVersion
 	config.GroupVersion = &gv
 	config.APIPath = "/apis"
-	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply cmd/apprepository-controller/pkg/client/clientset/versioned/typed/apprepository/v1alpha1/apprepository_client.go]: Most of the modifications in [that commit](https://github.com/kubeapps/kubeapps/pull/1359/commits/1e1a3ba7c4c3c7b0a177adcffec07169847c1a84) were suggested by @andresmgot so that we could import `helm.sh/helm/v3` at all.

**Comment by SimonAlling** [comment]
> Hmm, did you see the CI error? `FAIL use-cases/default-deployment.js (71.032s)`. Not sure if it was an intermittent error, or whether that is reproduced locally? (I mean, nothing should be effected since CI is running without the `useHelm3` flag, but haven't dug).

Yep, I saw it. No idea why, really. This line seems to be the one failing:

```javascript
await expect(page).toMatch("Ready", { timeout: 60000 });
```

**Comment by absoludity** [comment]
> > Hmm, did you see the CI error? `FAIL use-cases/default-deployment.js (71.032s)`. Not sure if it was an intermittent error, or whether that is reproduced locally? (I mean, nothing should be effected since CI is running without the `useHelm3` flag, but haven't dug).
> 
> Yep, I saw it. No idea why, really. This line seems to be the one failing:
> 
> ```js
> await expect(page).toMatch("Ready", { timeout: 60000 });
> ```

Hrm, testing this locally, I see the following when I try a default deploy of aerospike (well, anything):

e2e-failure

which would explain the failing test. I'll try to find if one of the updates has altered the token authentication which is done in tiller-proxy so you can move on.

I've done some sample tests you can merge for `agent/ListReleases` in https://github.com/SimonAlling/kubeapps/pull/1 .

File: pkg/chart/chart.go [hunk 5]
```diff
@@ -320,7 +320,7 @@ func (c *Chart) InitNetClient(details *Details) (HTTPClient, error) {
 
 	defaultHeaders := http.Header{"User-Agent": []string{c.userAgent}}
 	if auth.Header != nil {
-		secret, err := c.kubeClient.Core().Secrets(namespace).Get(auth.Header.SecretKeyRef.Name, metav1.GetOptions{})
+		secret, err := c.kubeClient.CoreV1().Secrets(namespace).Get(auth.Header.SecretKeyRef.Name, metav1.GetOptions{})
```
Comment (andresmgot) [inline pkg/chart/chart.go]: why this change?

**Comment by SimonAlling** [comment]
I just tried deploying Aerospike as well, and I wasn't able to reproduce the issue. I trust you and CircleCI more than myself, though.

**Comment by SimonAlling** [comment]
I think it would be interesting to run 1e1a3ba7c4c3c7b0a177adcffec07169847c1a84 through CI, because it's the only commit on this branch that I'd reasonably expect to affect any existing functionality.

**Comment by andresmgot** [comment]
I was able to also reproduce the issue (I just built the `tiller-proxy` image in this branch). The issue is related to the upgrade of `client-go`. I am not sure why but here:

https://github.com/kubeapps/kubeapps/blob/master/pkg/auth/auth.go#L77

The response returned is empty so both allowed and denied are `false`:

```
Status:SubjectAccessReviewStatus{Allowed:false,Reason:,EvaluationError:,Denied:false,}
```

If you are not able to reproduce the issue, it may be related to an issue with `kind` (which is the dev environment in the CI).

File: chart/kubeapps/values.yaml [hunk 6]
```diff
@@ -269,6 +269,28 @@ hooks:
...
   tolerations: {}
 
+# Kubeops is an interface between the Kubeapps Dashboard and Helm 3/Kubernetes.
+# Set useHelm3 to true to use Kubeops instead of Tiller Proxy.
+kubeops:
+  replicaCount: 2
+  image:
+    # TODO: These values are for development only!
+    registry: kubeapps
+    repository: kubeops
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply chart/kubeapps/values.yaml]: Yes, but that won't work for development, will it? See the nearby `TODO` comment.

File: cmd/kubeops/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+COPY cmd cmd
+ARG VERSION
+# With the trick below, Go's build cache is kept between builds.
+# https://github.com/golang/go/issues/27719#issuecomment-514747274
+RUN --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.cache/go-build \
+    CGO_ENABLED=0 go build -installsuffix cgo -ldflags "-X main.version=$VERSION" ./cmd/kubeops
+
+FROM scratch
+COPY --from=builder /go/src/github.com/kubeapps/kubeapps/kubeops /agent
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply cmd/kubeops/Dockerfile]: No, `agent` was just analogous to `proxy`. I'll change the name to `kubeops`.

File: cmd/kubeops/main.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	settings         environment.EnvSettings
+	chartsvcURL      string
+	helmDriverArg    string
+	userAgentComment string
+	listLimit        int
+	timeout          int64
+)
+
+func init() {
+	settings.AddFlags(pflag.CommandLine) // necessary???
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply cmd/kubeops/main.go]: More like I hoped that someone, somehow, would confirm the necessity of those lines. :sweat_smile:

File: chart/kubeapps/values.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply chart/kubeapps/values.yaml]: it will, once this reaches master and the kubeops image is built

File: cmd/kubeops/main.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply cmd/kubeops/main.go]: yes, that is adding the flags to the command

**Comment by andresmgot** [comment]
> If you are not able to reproduce the issue, it may be related to an issue with `kind` (which is the dev environment in the CI).

I am also able to reproduce it with minikube so I guess it's a different thing (btw, using Kubernetes 1.16)

File: pkg/agent/agent.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	case ConfigMap:
+		d := driver.NewConfigMaps(clientset.CoreV1().ConfigMaps(namespace))
+		d.Log = klog.Infof
+		store = storage.Init(d)
+	case Memory:
+		d := driver.NewMemory()
+		store = storage.Init(d)
+	default:
+		// No (real) enums/ADTs in Go, so no static guarantee against this case.
+		panic("Invalid Helm driver type: " + driverType)
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply pkg/agent/agent.go]: In this specific case, I'm actually not sure about that. I wanted to have the first argument to `agent.createStorage` be of a sum type, so that I'd only need to handle exactly the cases represented by that sum type. There seems to be no sum types in Go, so said first argument instead has the quasi sum type `DriverType`. This at least means that values like `os.Getenv("LOL"))` don't typecheck:

```
cannot use os.Getenv("LOL") (type string) as type agent.DriverType in argument to handler.WithAgentConfig
```

However, values like `"hello"` apparently count as members of `DriverType`, but I found no way around that.

I think returning an error in `createStorage` would introduce a lot of (IMO) unnecessary complexity in `agent.NewActionConfig` and `handler.WithAgentConfig`. I also think a programmer will understand that they're not supposed to pass `"hello"`, but an actual `DriverType` (e.g. `agent.Secret`).

At some point, you want to be able to say "I have a value of type _T_ – I know it has been validated _because_ it is a _T_." [_Parse, don't validate._](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) Go doesn't seem to help us more than I've already made it do.

I'm much more inclined to make `agent.ParseDriverType` return an error instead of panicking, because, unlike `createStorage`, it promises that it can handle any `string`. Its current deceitfulness stems from my interpretation of @absoludity's [suggestion](https://github.com/kubeapps/kubeapps/pull/1309#discussion_r349377228) that Kubeops should refuse to start if an invalid driver type has been specified. I'll be more than happy to return an error instead.

File: pkg/chart/chart.go (hunk 5 above)
Comment (SimonAlling) [reply pkg/chart/chart.go]: I don't know. I think you suggested it. (Do note that it's part of 1e1a3ba7c4c3c7b0a177adcffec07169847c1a84.)

File: cmd/kubeops/main.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	driverType := defaultHelmDriver
+	if helmDriverArg != "" {
+		driverType = agent.ParseDriverType(helmDriverArg)
+	} else {
+		// CLI argument was not provided; check environment variable.
+		helmDriverEnv := os.Getenv(driverEnvVar)
+		if helmDriverEnv != "" {
+			driverType = agent.ParseDriverType(helmDriverEnv)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply cmd/kubeops/main.go]: We did that because [Helm did](https://github.com/helm/helm/blob/f1dc84773f9a34fe59a504fdb32428ce1d56a2e8/cmd/helm/helm.go#L70). I'll be happy to remove the env var if you prefer just the flag.

File: pkg/agent/agent.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+	config.BearerToken = token
+	config.BearerTokenFile = ""
+	clientset, err := kubernetes.NewForConfig(config)
+	store := createStorage(driver, namespace, clientset)
+	actionConfig.RESTClientGetter = nil     // TODO replace nil with meaningful value
+	actionConfig.KubeClient = kube.New(nil) // TODO replace nil with meaningful value
```
[the remark this answers is not in the document]
Comment (SimonAlling) [reply pkg/agent/agent.go]: I guess it depends on your definition of "necessary". We probably thought it would be nice to be explicit. What would you recommend: keep or remove?

**Comment by andresmgot** [comment]
Finally found the issue (this one was a tough one). The problem is that the `BearerTokenFile` was set when creating the `config` in the `auth` module and now, it's taking precedence over the `BearerToken`. Because of that, it's using the default serviceaccount to validate all the actions. To fix the issue, add the line:

```patch
--- a/pkg/auth/auth.go
+++ b/pkg/auth/auth.go
@@ -105,6 +105,7 @@ func NewAuth(token string) (*UserAuth, error) {
        }
        // Overwrite default token
        config.BearerToken = token
+       config.BearerTokenFile = ""
        kubeClient, err := kubernetes.NewForConfig(config)
        if err != nil {
                return nil, err
```

here: https://github.com/kubeapps/kubeapps/blob/master/pkg/auth/auth.go#L107

File: pkg/agent/agent.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/agent/agent.go]: it's okay to leave it, I guess what I don't understand is the `TODO` note, why should it have a meaningful value? What's a meaningful value in this context?

File: pkg/agent/agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/agent/agent.go]: If I understand this correctly, you mean that `driverType` was validated before so this won't never panic, right? (because the driver was parsed before). If that's the case then it's fine, I just found it weird for a function in a library to panic.

File: pkg/chart/chart.go (hunk 5 above)
Comment (andresmgot) [reply pkg/chart/chart.go]: I guess that was me trying things, you don't need this change

**Comment by SimonAlling** [comment]
> It's okay for me to have a starting point. Leaving for @absoludity to do a final review. Thanks!

Thank you so much for helping me out with the debugging! :smiley:

File: pkg/agent/agent.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (SimonAlling) [reply pkg/agent/agent.go]: I don't know; I still don't fully grasp how everything fits together in this fairly complex system. I thought `nil` was obviously not meaningful and would need to be replaced sooner or later.

File: pkg/agent/agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (SimonAlling) [reply pkg/agent/agent.go]: Yeah, it is weird. But Go doesn't seem to let me express a good old sum type, so this was the best solution I could come up with. It is correct that `createStorage` should never panic, because `driverType` should really be a `DriverType`. The only way I know to violate this assumption is to pass a string _literal_ as `driverType`.
